File: src/main/java/io/github/boogiemonster1o1/cartses/entity/cart/MinecartWithBarrelEntity.java
package io.github.boogiemonster1o1.cartses.entity.cart;

import com.chocohead.mm.api.ClassTinkerers;
import io.github.boogiemonster1o1.cartses.networking.EntityPacketUtils;

import net.minecraft.block.BarrelBlock;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.damage.DamageSource;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.entity.vehicle.StorageMinecartEntity;
import net.minecraft.network.Packet;
import net.minecraft.screen.GenericContainerScreenHandler;
import net.minecraft.screen.ScreenHandler;
import net.minecraft.util.math.Direction;
import net.minecraft.world.GameRules;
import net.minecraft.world.World;

@SuppressWarnings("EntityConstructor")
public class MinecartWithBarrelEntity extends StorageMinecartEntity {
	public static final Type MINECART_TYPE = ClassTinkerers.getEnum(Type.class, "BARREL");

	public MinecartWithBarrelEntity(EntityType<?> entityType, World world) {
		super(entityType, world);
	}

	public MinecartWithBarrelEntity(EntityType<?> type, double x, double y, double z, World world) {
		super(type, x, y, z, world);
	}

	@Override
	protected ScreenHandler getScreenHandler(int syncId, PlayerInventory playerInventory) {
		return GenericContainerScreenHandler.createGeneric9x3(syncId, playerInventory, this);
	}

	@Override
	public Type getMinecartType() {
		return MINECART_TYPE;
	}

	@Override
	public Packet<?> createSpawnPacket() {
		return EntityPacketUtils.createPacket(this);
	}

	@Override
	public int size() {
		return 27;
	}

	@Override
	public int getDefaultBlockOffset() {
		return 8;
	}

	@Override
	public BlockState getDefaultContainedBlock() {
		return Blocks.BARREL.getDefaultState().with(BarrelBlock.FACING, Direction.UP);
	}

	@Override
	public void dropItems(DamageSource damageSource) {
		super.dropItems(damageSource);
		if (this.world.getGameRules().getBoolean(GameRules.DO_ENTITY_DROPS)) {
			this.dropItem(Blocks.BARREL);
		}
	}
}

File: src/main/scala/io/github/boogiemonster1o1/cartses/entity/cart/MinecartWithBarrelEntity.scala
package io.github.boogiemonster1o1.cartses.entity.cart

import io.github.boogiemonster1o1.cartses.entity.MinecartTypes
import io.github.boogiemonster1o1.cartses.entity.networking.EntityUtils
import net.minecraft.block.{BarrelBlock, BlockState, Blocks}
import net.minecraft.entity.EntityType
import net.minecraft.entity.damage.DamageSource
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.entity.vehicle.{AbstractMinecartEntity, StorageMinecartEntity}
import net.minecraft.network.Packet
import net.minecraft.screen.{GenericContainerScreenHandler, ScreenHandler}
import net.minecraft.util.math.{Direction, Vec3d}
import net.minecraft.world.{GameRules, World}

@SuppressWarnings(Array("EntityConstructor")) class MinecartWithBarrelEntity(entityType: EntityType[_], world: World) extends StorageMinecartEntity(entityType, world) {
  	def this(`type`: EntityType[_], x: Double, y: Double, z: Double, world: World) {
		this(`type`, world)
		EntityUtils.setupPos(this, x, y, z)
	}

	override protected def getScreenHandler(syncId: Int, playerInventory: PlayerInventory): ScreenHandler = GenericContainerScreenHandler.createGeneric9x3(syncId, playerInventory, this)

	override def getMinecartType: AbstractMinecartEntity.Type = MinecartTypes.barrel

	override def createSpawnPacket: Packet[_] = EntityUtils.createPacket(this)

	override def size = 27

	override def getDefaultBlockOffset = 8

	override def getDefaultContainedBlock: BlockState = Blocks.BARREL.getDefaultState.`with`(BarrelBlock.FACING, Direction.UP)

	override def dropItems(damageSource: DamageSource): Unit = {
		super.dropItems(damageSource)
		if (this.world.getGameRules.getBoolean(GameRules.DO_ENTITY_DROPS)) this.dropItem(Blocks.BARREL)
	}
}
